### PR TITLE
[OPPRO-233] Add SingularOrList support

### PR DESCRIPTION
**Pull request by jinchengchenghh**

### CODE REVIEW

**Comment by jinchengchenghh** [comment]
@rui-mo @zhejiangxiaomai Can you help review? Thanks!

**Comment by rui-mo** [comment]
Thanks, I'm working on fixing the issue you met when running a query like below, and will refactor a bit based on your change for `in`.

```
select l_orderkey from lineitem where l_orderkey > 2  and l_orderkey = 1
```

File: velox/substrait/SubstraitToVeloxPlan.cpp [hunk 1]
```diff
@@ -1594,4 +1596,106 @@ SubstraitVeloxPlanConverter::connectWithAnd(
...
+  ::substrait::Expression_FieldReference selection;
+  if (singularOrList.value().has_scalar_function()) {
+    selection = getExprFromFunctionArgument(
+                    singularOrList.value().scalar_function().arguments()[0])
+                    .selection();
+  } else if (singularOrList.value().has_selection()) {
+    selection = singularOrList.value().selection();
+  }
+  return subParser_->parseReferenceSegment(selection.direct_reference());
+  ;
```
Comment (zhejiangxiaomai) [inline velox/substrait/SubstraitToVeloxPlan.cpp:1666]: remove ;

File: velox/substrait/SubstraitToVeloxPlan.cpp [hunk 2]
```diff
@@ -480,9 +495,13 @@ std::shared_ptr<const core::PlanNode> SubstraitVeloxPlanConverter::toVeloxPlan(
       subfieldFilters.clear();
       remainingFilter =
```
Comment (zhejiangxiaomai) [inline velox/substrait/SubstraitToVeloxPlan.cpp:496]: This variable “remainingFilter”  will not used.

File: velox/substrait/SubstraitToVeloxPlan.cpp [hunk 3]
```diff
@@ -480,9 +495,13 @@ std::shared_ptr<const core::PlanNode> SubstraitVeloxPlanConverter::toVeloxPlan(
...
           connectWithAnd(colNameList, veloxTypeList, scalarFunctions);
+      remainingFilter = connectWithAnd(
+          remainingFilter, colNameList, veloxTypeList, singularOrLists);
     } else {
       remainingFilter =
```
Comment (zhejiangxiaomai) [inline velox/substrait/SubstraitToVeloxPlan.cpp:501]: This variable “remainingFilter” will not used.

Comment (jinchengchenghh) [reply velox/substrait/SubstraitToVeloxPlan.cpp:501]: it used here https://github.com/oap-project/velox/pull/45/files#diff-6bbec4a6c456cb1efcf0c8f2dae7b80982e9002b9cf2a3460fd1ae545f291e49R512

File: velox/substrait/SubstraitToVeloxPlan.cpp [hunk 4]
```diff
@@ -1021,50 +1040,67 @@ bool SubstraitVeloxPlanConverter::canPushdownOr(
...
-      sIsNotNull, sGte, sGt, sLte, sLt, sEqual, sIn};
+      sIsNotNull, sGte, sGt, sLte, sLt, sEqual};
 
   bool inExists = false;
   for (const auto& arg : scalarFunction.arguments()) {
-    if (!getExprFromFunctionArgument(arg).has_scalar_function()) {
+    if (getExprFromFunctionArgument(arg).has_scalar_function()) {
       // Or relation betweeen literals is not supported to be pushded down
       // currently.
       return false;
```
Comment (rui-mo) [inline velox/substrait/SubstraitToVeloxPlan.cpp:1050]: Seems to be not correct to just return false here. Will fix this locally.

File: velox/substrait/SubstraitToVeloxPlan.cpp [hunk 5]
```diff
@@ -1111,51 +1149,16 @@ void SubstraitVeloxPlanConverter::separateFilters(
...
-                       .list();
-  variants.reserve(valueList.values().size());
-  for (const auto& literal : valueList.values()) {
-    variants.emplace_back(
-        exprConverter_->toTypedVariant(literal)->veloxVariant);
+bool SubstraitVeloxPlanConverter::canPushdownSingularList(
+    const ::substrait::Expression_SingularOrList& singularList) {
+  if (singularList.value().has_selection()) {
+    return true;
+  } else if (singularList.value().has_scalar_function()) {
```
Comment (rui-mo) [inline velox/substrait/SubstraitToVeloxPlan.cpp:1156]: When `value` is not field, the IN filter cannot be pushed down. Will fix this in a following PR.

**Pull request merged**